Doctrine/Orm/Extension/PaginationExtension.php

```diff
@@ -129,4 +142,229 @@ private function getItemsPerPage(ResourceInterface $resource, Request $request)
 
         return $resource->getItemsPerPageByDefault();
     }
+
+    /**
+     * Determines whether output walkers should be used.
+     *
+     * @param QueryBuilder $queryBuilder
+     *
+     * @return bool
+     */
+    private function useOutputWalkers(QueryBuilder $queryBuilder)
+    {
+        // Disable output walkers by default (performance)
+        $useOutputWalkers = false;
+
+        /*
+         * "Cannot count query that uses a HAVING clause. Use the output walkers for pagination"
+         *
+         * @see https://github.com/doctrine/doctrine2/blob/900b55d16afdcdeb5100d435a7166d3a425b9873/lib/Doctrine/ORM/Tools/Pagination/CountWalker.php#L50
+         */
+        if (!empty($queryBuilder->getDQLPart('having'))) {
+            $useOutputWalkers = true;
+        }
+
+        /*
+         * "Paginating an entity with foreign key as identifier only works when using the Output Walkers. Call Paginator#setUseOutputWalkers(true) before iterating the paginator."
+         *
+         * @see https://github.com/doctrine/doctrine2/blob/900b55d16afdcdeb5100d435a7166d3a425b9873/lib/Doctrine/ORM/Tools/Pagination/LimitSubqueryWalker.php#L87
+         */
+        foreach ($queryBuilder->getRootEntities() as $rootEntity) {
```

Comment:
Good suggestion but wouldn't it be overkill in this case? (And it only leads to more cryptic/confusing method names...)